File: examples/handler.php
<?php

declare(strict_types=1);

/**
 * Example roxy upstream handler (FastCGI / HTTP).
 *
 * roxy translates MCP JSON-RPC into a simple envelope and POSTs it to this
 * script. The handler inspects `type` and returns a JSON payload that roxy
 * turns back into an MCP response.
 *
 * All array shapes below use PHPStan / Psalm syntax (`array{key: T, ...}`) so
 * static analyzers can verify both the decoded request and the encoded reply.
 *
 * See `src/protocol.rs` (`UpstreamEnvelope`, `UpstreamRequest`,
 * `UpstreamCallResult`) for the authoritative Rust-side contract.
 */

// Prefer the CGI/FastCGI body stream (`php://input`); fall back to STDIN so the
// script can also be driven from the shell for local debugging.
$input = file_get_contents('php://input');
if ($input === '' || $input === false) {
    $input = file_get_contents('php://stdin');
}

/**
 * Decoded envelope sent by roxy. Every request carries the common envelope
 * fields (`session_id`, `request_id`, `type`) plus fields specific to the
 * discriminated `type` variant.
 *
 * @var array{
 *     type?: 'discover'|'call_tool'|'read_resource'|'get_prompt'|'elicitation_cancelled',
 *     session_id?: string|null,  // MCP session id, null under stdio transport
 *     request_id?: string,       // opaque id used by roxy to correlate replies
 *     name?: string,             // tool / prompt name (call_tool, get_prompt, elicitation_cancelled)
 *     arguments?: array<string, mixed>,   // tool / prompt arguments
 *     elicitation_results?: list<array<string, mixed>>, // prior elicitation replies, oldest first
 *     context?: mixed,           // opaque value previously returned in an `elicit` reply
 *     uri?: string,              // resource URI (read_resource)
 *     action?: string,           // elicitation action: 'decline' | 'cancel'
 * }|null|false $request
 */
$request = json_decode($input, true);

header('Content-Type: application/json');

if (!$request || !isset($request['type'])) {
    echo json_encode([
        // Error envelope — roxy maps this to a JSON-RPC error back to the client.
        'error' => [
            'code' => 400,               // integer status, mirrored into JSON-RPC `code`
            'message' => 'Invalid request: missing type field', // human-readable detail
        ],
    ]);
    exit;
}

// Envelope fields present on every request. Kept around for logging / audit;
// the example handlers below don't use them but real deployments typically do.
$sessionId = $request['session_id'] ?? null;
$requestId = $request['request_id'] ?? null;

echo match ($request['type']) {
    'discover'              => handleDiscover(),
    'call_tool'             => handleCallTool($request),
    'read_resource'         => handleReadResource($request),
    'get_prompt'            => handleGetPrompt($request),
    'elicitation_cancelled' => handleElicitationCancelled($request),
    default => json_encode([
        'error' => [
            'code' => 400,
            'message' => "Unknown request type: {$request['type']}",
        ],
    ]),
};

/**
 * Answer a `discover` request: advertise every tool, resource and prompt this
 * backend exposes. roxy caches the result at startup and uses it to serve
 * `tools/list`, `resources/list` and `prompts/list` MCP methods without
 * calling the backend again.
 *
 * @return string JSON-encoded {@see UpstreamDiscoverResponse} payload.
 */
function handleDiscover(): string
{
    return json_encode([
        /**
         * Tool catalog.
         *
         * @var list<array{
         *     name: string,               // stable machine id — used by call_tool
         *     title?: string,              // human label shown by MCP clients
         *     description?: string,        // free-form description shown by clients
         *     input_schema?: array<string, mixed>,  // JSON Schema for `arguments`
         *     output_schema?: array<string, mixed>, // optional JSON Schema for `structured_content`
         * }>
         */
        'tools' => [
            [
                'name'        => 'echo',                  // tool id — must be unique within this backend
                'title'       => 'Echo Message',          // label displayed in MCP clients
                'description' => 'Echoes back the input message', // shown next to the title
                'input_schema' => [
                    'type'       => 'object',             // JSON Schema root type
                    'properties' => [
                        'message' => [
                            'type'        => 'string',    // JSON Schema type for this field
                            'description' => 'The message to echo', // field-level hint
                        ],
                    ],
                    'required' => ['message'],            // list of keys that MUST be present
                ],
            ],
            [
                'name'        => 'add',
                'title'       => 'Add Numbers',
                'description' => 'Adds two numbers and returns structured result',
                'input_schema' => [
                    'type'       => 'object',
                    'properties' => [
                        // `number` covers both integer and float in JSON Schema.
                        'a' => ['type' => 'number', 'description' => 'First number'],
                        'b' => ['type' => 'number', 'description' => 'Second number'],
                    ],
                    'required' => ['a', 'b'],
                ],
                // Optional — when set, MCP clients can validate `structured_content`
                // and render it with field names / types.
                'output_schema' => [
                    'type'       => 'object',
                    'properties' => [
                        'sum'      => ['type' => 'number'], // the computed sum
                        'operands' => [                     // echoes the inputs for audit
                            'type'       => 'object',
                            'properties' => [
                                'a' => ['type' => 'number'],
                                'b' => ['type' => 'number'],
                            ],
                        ],
                    ],
                ],
            ],
            [
                'name'        => 'book_flight',
                'title'       => 'Book a Flight',
                'description' => 'Books a flight with elicitation for missing details',
                'input_schema' => [
                    'type'       => 'object',
                    'properties' => [
                        'destination' => [
                            'type'        => 'string',
                            'description' => 'Flight destination',
                        ],
                    ],
                    'required' => ['destination'],
                ],
            ],
        ],
        /**
         * Resource catalog.
         *
         * @var list<array{
         *     uri: string,           // identifier clients pass back in `read_resource`
         *     name: string,          // stable machine id
         *     title?: string,        // human label
         *     description?: string,  // free-form description
         *     mime_type?: string,    // IANA mime type of the resource body
         * }>
         */
        'resources' => [
            [
                'uri'         => 'roxy://status',        // opaque, scheme is up to the backend
                'name'        => 'server-status',         // stable machine id
                'title'       => 'Server Status',         // shown by clients
                'description' => 'Current server status', // shown by clients
                'mime_type'   => 'application/json',      // tells the client how to render the body
            ],
        ],
        /**
         * Prompt catalog.
         *
         * @var list<array{
         *     name: string,          // stable machine id — used by get_prompt
         *     title?: string,        // human label
         *     description?: string,  // free-form description
         *     arguments?: list<array{
         *         name: string,          // argument name
         *         title?: string,        // human label
         *         description?: string,  // free-form description
         *         required?: bool,       // defaults to false
         *     }>,
         * }>
         */
        'prompts' => [
            [
                'name'        => 'greet',
                'title'       => 'Greeting',
                'description' => 'Generate a greeting',
                'arguments' => [
                    [
                        'name'        => 'name',         // argument key passed in `get_prompt.arguments`
                        'title'       => 'Person Name',  // human label
                        'description' => 'Name to greet',
                        'required'    => true,           // MCP clients will prompt the user before sending
                    ],
                ],
            ],
        ],
    ]);
}

/**
 * Dispatch a `call_tool` request to the matching handler.
 *
 * @param array{
 *     name?: string,
 *     arguments?: array<string, mixed>,
 *     elicitation_results?: list<array<string, mixed>>,
 *     context?: mixed,
 * } $request Decoded envelope fields for the `call_tool` variant.
 * @return string JSON-encoded {@see UpstreamCallResult} payload.
 */
function handleCallTool(array $request): string
{
    $name                = $request['name']                ?? '';
    $args                = $request['arguments']           ?? [];
    // Ordered list of previous elicitation replies from the client, oldest first.
    // Empty on the first call; populated on each follow-up while a multi-step
    // elicitation flow is in progress.
    $elicitationResults  = $request['elicitation_results'] ?? [];
    // Opaque value that this handler returned inside a previous `elicit` reply.
    // roxy threads it back verbatim so the handler can restore its own state.
    $context             = $request['context']             ?? null;

    return match ($name) {
        'echo' => json_encode([
            /**
             * Content blocks are the user-visible part of the reply.
             *
             * @var list<array{type: 'text', text: string}|array{
             *     type: 'resource_link',
             *     uri: string,
             *     name: string,
             *     title?: string,
             *     description?: string,
             *     mime_type?: string,
             * }>
             */
            'content' => [
                [
                    'type' => 'text',                         // discriminator for the content block kind
                    'text' => $args['message'] ?? '',         // the string to display
                ],
            ],
        ]),
        'add'         => handleAdd($args),
        'book_flight' => handleBookFlight($args, $elicitationResults, $context),
        default       => json_encode([
            'error' => [
                'code'    => 404,
                'message' => "Unknown tool: {$name}",
            ],
        ]),
    };
}

/**
 * Implementation of the `add` tool.
 *
 * @param array{a?: int|float, b?: int|float} $args Tool arguments as declared by `input_schema`.
 * @return string JSON-encoded content + `structured_content` reply.
 */
function handleAdd(array $args): string
{
    $a   = $args['a'] ?? 0;
    $b   = $args['b'] ?? 0;
    $sum = $a + $b;

    return json_encode([
        // Human-readable rendering of the result.
        'content' => [
            [
                'type' => 'text',
                'text' => "{$a} + {$b} = {$sum}",
            ],
        ],
        /**
         * Machine-readable result, validated against the tool's `output_schema`
         * by MCP clients that support structured output.
         *
         * @var array{sum: int|float, operands: array{a: int|float, b: int|float}}
         */
        'structured_content' => [
            'sum'      => $sum,                       // matches output_schema.properties.sum
            'operands' => ['a' => $a, 'b' => $b],     // echoes the inputs for audit
        ],
    ]);
}

/**
 * Multi-turn example: first call returns an `elicit` asking for flight class;
 * second call (with the elicitation reply) completes the booking.
 *
 * @param array{destination?: string}                  $args               Tool arguments.
 * @param list<array<string, mixed>>                   $elicitationResults Prior user replies.
 * @param mixed                                         $context            Opaque value returned in the previous `elicit`.
 * @return string JSON-encoded {@see UpstreamCallResult} payload.
 */
function handleBookFlight(array $args, array $elicitationResults, mixed $context): string
{
    $destination = $args['destination'] ?? 'Unknown';

    // First round: no prior elicitation replies — ask the client for the class.
    if (empty($elicitationResults)) {
        return json_encode([
            /**
             * Elicit reply — roxy forwards it to the MCP client as an
             * `elicitation/create` request. The client asks the user and then
             * calls this same tool again with `elicitation_results` populated.
             *
             * @var array{
             *     message: string,
             *     schema: array<string, mixed>,
             *     context?: mixed,
             * }
             */
            'elicit' => [
                'message' => "Select flight class for {$destination}", // prompt shown to the user
                // JSON Schema describing the user's expected reply. Clients
                // render the fields as a form.
                'schema' => [
                    'type'       => 'object',
                    'properties' => [
                        'class' => [
                            'type'  => 'string',
                            'title' => 'Flight Class',
                            'enum'  => ['economy', 'business', 'first'], // allowed values
                        ],
                    ],
                    'required' => ['class'],
                ],
                // Opaque handler-side bookkeeping. roxy stores it and replays
                // it to us on the follow-up call. Use it for whatever you need
                // to resume — here it records which step of the flow we're on.
                'context' => [
                    'destination' => $destination,
                    'step'        => 1,
                ],
            ],
        ]);
    }

    // Second round: the reply carries the class the user picked.
    $class     = $elicitationResults[0]['class'] ?? 'economy';
    $bookingId = rand(1000, 9999);

    return json_encode([
        'content' => [
            // Plain-text summary.
            [
                'type' => 'text',
                'text' => "Booked {$class} flight to {$destination}. Booking #{$bookingId}",
            ],
            // A `resource_link` lets the client fetch details via `read_resource`.
            [
                'type'  => 'resource_link',                         // discriminator
                'uri'   => "roxy://bookings/{$bookingId}",          // target of the follow-up read_resource
                'name'  => "booking-{$bookingId}",                  // stable machine id
                'title' => "Booking #{$bookingId}",                 // human label
            ],
        ],
        /**
         * Structured equivalent of the booking for programmatic consumers.
         *
         * @var array{booking_id: int, destination: string, class: string}
         */
        'structured_content' => [
            'booking_id'  => $bookingId,   // generated booking id
            'destination' => $destination, // echoed from the original arguments
            'class'       => $class,       // value picked by the user during elicitation
        ],
    ]);
}

/**
 * Answer a `read_resource` request.
 *
 * @param array{uri?: string} $request Decoded envelope for the read_resource variant.
 * @return string JSON-encoded content reply or error payload.
 */
function handleReadResource(array $request): string
{
    $uri = $request['uri'] ?? '';

    if ($uri === 'roxy://status') {
        return json_encode([
            // Resource bodies ride in the same `content` array as tool output.
            'content' => [
                [
                    'type' => 'text',
                    // Stringified JSON — declared as `application/json` in discover.
                    'text' => json_encode([
                        'status' => 'running',    // fixed in this demo
                        'php'    => PHP_VERSION,  // observed at runtime
                    ]),
                ],
            ],
        ]);
    }

    return json_encode([
        'error' => [
            'code'    => 404,
            'message' => "Unknown resource: {$uri}",
        ],
    ]);
}

/**
 * Answer a `get_prompt` request.
 *
 * @param array{name?: string, arguments?: array<string, mixed>} $request
 * @return string JSON-encoded content reply or error payload.
 */
function handleGetPrompt(array $request): string
{
    $name = $request['name']      ?? '';
    $args = $request['arguments'] ?? [];

    if ($name === 'greet') {
        $personName = $args['name'] ?? 'World';
        return json_encode([
            'content' => [
                [
                    'type' => 'text',
                    'text' => "Hello, {$personName}! How can I help you today?",
                ],
            ],
        ]);
    }

    return json_encode([
        'error' => [
            'code'    => 404,
            'message' => "Unknown prompt: {$name}",
        ],
    ]);
}

/**
 * Notification that the user declined or cancelled a pending elicitation.
 * roxy doesn't wait for a meaningful body — it just wants the handler to know
 * so any server-side state tied to the flow can be cleaned up.
 *
 * @param array{name?: string, action?: string, context?: mixed} $request
 * @return string JSON-encoded acknowledgement `{ok: true}`.
 */
function handleElicitationCancelled(array $request): string
{
    $name   = $request['name']   ?? '';  // tool the elicitation belonged to
    $action = $request['action'] ?? '';  // 'decline' or 'cancel'
    error_log("Elicitation {$action} for tool: {$name}");

    return json_encode(['ok' => true]);
}

File: examples/handler.py
#!/usr/bin/env python3
"""
Example roxy upstream handler (HTTP transport, Python 3 stdlib only).

roxy translates MCP JSON-RPC into a simple envelope and POSTs it to this
server. The handler inspects ``type`` and returns a JSON payload that roxy
turns back into an MCP response.

See ``src/protocol.rs`` (``UpstreamEnvelope``, ``UpstreamRequest``,
``UpstreamCallResult``) for the authoritative Rust-side contract, and
``examples/handler.php`` / ``examples/handler.ts`` for the functionally-
identical ports in other languages.

Run::

    python3 examples/handler.py               # listens on :8000
    PORT=9001 python3 examples/handler.py     # override port

Point roxy at it::

    cargo run -- --upstream http://127.0.0.1:8000/

No third-party dependencies — only the Python 3.9+ standard library.
"""

from __future__ import annotations

import json
import logging
import os
import random
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any, Dict, List, Union

# ---------------------------------------------------------------------------
# Protocol type aliases — mirror ``src/protocol.rs``.
#
# Python doesn't enforce these at runtime but the aliases document the exact
# shapes roxy sends and expects, matching the PHP (PHPStan) and TypeScript
# variants of this example.
# ---------------------------------------------------------------------------

#: Arbitrary JSON value (replacement for the recursive ``serde_json::Value``).
JsonValue = Union[
    None, bool, int, float, str, List[Any], Dict[str, Any]
]

#: Decoded envelope from roxy. Fields present depend on the ``type`` variant.
#:
#: Common fields:
#:   - ``type``       : one of ``discover`` | ``call_tool`` | ``read_resource``
#:                       | ``get_prompt`` | ``elicitation_cancelled``
#:   - ``session_id`` : MCP session id, ``None`` under stdio transport
#:   - ``request_id`` : opaque id roxy uses to correlate replies
#:
#: Variant-specific fields:
#:   - ``call_tool``             : ``name``, ``arguments``, ``elicitation_results``, ``context``
#:   - ``read_resource``         : ``uri``
#:   - ``get_prompt``            : ``name``, ``arguments``
#:   - ``elicitation_cancelled`` : ``name``, ``action``, ``context``
Request = Dict[str, Any]

#: Standard reply body shared by ``call_tool`` / ``read_resource`` / ``get_prompt``.
#:
#: Keys:
#:   - ``content``            : list of content blocks (see below), user-visible
#:   - ``structured_content`` : optional machine-readable payload validated
#:                              against the tool's ``output_schema``
#:
#: Alternatively a handler may return:
#:   - ``{"error": {"code": int, "message": str}}`` — mapped to JSON-RPC error
#:   - ``{"elicit": {"message": str, "schema": {...}, "context": <any>}}``
#:     — mapped to an MCP ``elicitation/create`` request
Reply = Dict[str, Any]


# ---------------------------------------------------------------------------
# Handlers
# ---------------------------------------------------------------------------


def handle_discover() -> Reply:
    """Advertise every tool, resource and prompt this backend exposes.

    roxy caches the result at startup and uses it to serve ``tools/list``,
    ``resources/list`` and ``prompts/list`` without re-calling us.
    """
    return {
        # Tool catalog. Each entry:
        #   name          : stable machine id — used by call_tool
        #   title         : human label shown by MCP clients
        #   description   : free-form description shown by clients
        #   input_schema  : JSON Schema for ``arguments``
        #   output_schema : optional JSON Schema for ``structured_content``
        "tools": [
            {
                "name": "echo",                                    # tool id — unique per backend
                "title": "Echo Message",                           # label in MCP clients
                "description": "Echoes back the input message",    # shown next to the title
                "input_schema": {
                    "type": "object",                              # JSON Schema root type
                    "properties": {
                        "message": {
                            "type": "string",                      # JSON Schema type
                            "description": "The message to echo",  # field-level hint
                        },
                    },
                    "required": ["message"],                       # keys that MUST be present
                },
            },
            {
                "name": "add",
                "title": "Add Numbers",
                "description": "Adds two numbers and returns structured result",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        # ``number`` covers both int and float in JSON Schema.
                        "a": {"type": "number", "description": "First number"},
                        "b": {"type": "number", "description": "Second number"},
                    },
                    "required": ["a", "b"],
                },
                # Optional — lets clients validate ``structured_content`` and
                # render it with field names / types.
                "output_schema": {
                    "type": "object",
                    "properties": {
                        "sum": {"type": "number"},   # the computed sum
                        "operands": {                # echoes the inputs for audit
                            "type": "object",
                            "properties": {
                                "a": {"type": "number"},
                                "b": {"type": "number"},
                            },
                        },
                    },
                },
            },
            {
                "name": "book_flight",
                "title": "Book a Flight",
                "description": "Books a flight with elicitation for missing details",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "destination": {
                            "type": "string",
                            "description": "Flight destination",
                        },
                    },
                    "required": ["destination"],
                },
            },
        ],
        # Resource catalog. Each entry:
        #   uri         : identifier clients pass back in ``read_resource``
        #   name        : stable machine id
        #   title       : human label
        #   description : free-form description
        #   mime_type   : IANA mime type of the resource body
        "resources": [
            {
                "uri": "roxy://status",             # opaque, scheme is up to the backend
                "name": "server-status",            # stable machine id
                "title": "Server Status",           # shown by clients
                "description": "Current server status",  # shown by clients
                "mime_type": "application/json",    # how the client should render the body
            },
        ],
        # Prompt catalog. Each entry:
        #   name        : stable machine id — used by get_prompt
        #   title       : human label
        #   description : free-form description
        #   arguments   : list of {name, title, description, required}
        "prompts": [
            {
                "name": "greet",
                "title": "Greeting",
                "description": "Generate a greeting",
                "arguments": [
                    {
                        "name": "name",                # argument key passed in get_prompt.arguments
                        "title": "Person Name",        # human label
                        "description": "Name to greet",
                        "required": True,              # clients prompt the user before sending
                    },
                ],
            },
        ],
    }


def handle_call_tool(request: Request) -> Reply:
    """Dispatch a ``call_tool`` request to the matching sub-handler."""
    name: str = request.get("name", "")
    args: Dict[str, Any] = request.get("arguments") or {}
    # Ordered list of previous elicitation replies, oldest first. Empty on the
    # first call; populated on each follow-up while a multi-step elicitation
    # flow is in progress.
    elicitation_results: List[Dict[str, Any]] = request.get("elicitation_results") or []
    # Opaque value this handler returned in a previous ``elicit`` reply. roxy
    # threads it back verbatim so the handler can restore its own state.
    context: Any = request.get("context")

    if name == "echo":
        message = args.get("message", "")
        return {
            "content": [
                {
                    "type": "text",                # discriminator for the content block kind
                    "text": str(message),          # the string to display
                },
            ],
        }

    if name == "add":
        return handle_add(args)

    if name == "book_flight":
        return handle_book_flight(args, elicitation_results, context)

    return {
        "error": {
            "code": 404,
            "message": f"Unknown tool: {name}",
        },
    }


def handle_add(args: Dict[str, Any]) -> Reply:
    """Implementation of the ``add`` tool."""
    a = args.get("a", 0)
    b = args.get("b", 0)
    total = a + b

    return {
        # Human-readable rendering of the result.
        "content": [
            {
                "type": "text",
                "text": f"{a} + {b} = {total}",
            },
        ],
        # Machine-readable result — matches the tool's ``output_schema``.
        "structured_content": {
            "sum": total,                     # matches output_schema.properties.sum
            "operands": {"a": a, "b": b},     # echoes the inputs for audit
        },
    }


def handle_book_flight(
    args: Dict[str, Any],
    elicitation_results: List[Dict[str, Any]],
    _context: Any,
) -> Reply:
    """Multi-turn example.

    First call returns an ``elicit`` asking for flight class; the second call
    (with the elicitation reply) completes the booking.
    """
    destination = args.get("destination", "Unknown")

    # First round: no prior elicitation replies — ask the client for the class.
    if not elicitation_results:
        return {
            "elicit": {
                "message": f"Select flight class for {destination}",  # prompt shown to user
                # JSON Schema describing the expected reply. Clients render
                # the fields as a form.
                "schema": {
                    "type": "object",
                    "properties": {
                        "class": {
                            "type": "string",
                            "title": "Flight Class",
                            "enum": ["economy", "business", "first"],  # allowed values
                        },
                    },
                    "required": ["class"],
                },
                # Opaque handler-side bookkeeping. roxy replays it on the
                # follow-up call. Use it for whatever you need to resume — here
                # it records which step of the flow we're on.
                "context": {
                    "destination": destination,
                    "step": 1,
                },
            },
        }

    # Second round: the reply carries the class the user picked.
    reply = elicitation_results[0] if elicitation_results else {}
    cls = reply.get("class", "economy")
    booking_id = random.randint(1000, 9999)

    return {
        "content": [
            # Plain-text summary.
            {
                "type": "text",
                "text": f"Booked {cls} flight to {destination}. Booking #{booking_id}",
            },
            # A ``resource_link`` lets the client fetch details via read_resource.
            {
                "type": "resource_link",                        # discriminator
                "uri": f"roxy://bookings/{booking_id}",          # target of the follow-up read_resource
                "name": f"booking-{booking_id}",                 # stable machine id
                "title": f"Booking #{booking_id}",               # human label
            },
        ],
        "structured_content": {
            "booking_id": booking_id,    # generated booking id
            "destination": destination,  # echoed from the original arguments
            "class": cls,                # value picked by the user during elicitation
        },
    }


def handle_read_resource(request: Request) -> Reply:
    """Answer a ``read_resource`` request."""
    uri: str = request.get("uri", "")

    if uri == "roxy://status":
        return {
            # Resource bodies ride in the same ``content`` array as tool output.
            "content": [
                {
                    "type": "text",
                    # Stringified JSON — declared as application/json in discover.
                    "text": json.dumps(
                        {
                            "status": "running",       # fixed in this demo
                            "python": sys.version.split()[0],  # observed at runtime
                        }
                    ),
                },
            ],
        }

    return {
        "error": {
            "code": 404,
            "message": f"Unknown resource: {uri}",
        },
    }


def handle_get_prompt(request: Request) -> Reply:
    """Answer a ``get_prompt`` request."""
    name: str = request.get("name", "")
    args: Dict[str, Any] = request.get("arguments") or {}

    if name == "greet":
        person_name = args.get("name", "World")
        return {
            "content": [
                {
                    "type": "text",
                    "text": f"Hello, {person_name}! How can I help you today?",
                },
            ],
        }

    return {
        "error": {
            "code": 404,
            "message": f"Unknown prompt: {name}",
        },
    }


def handle_elicitation_cancelled(request: Request) -> Reply:
    """Notification that the user declined or cancelled a pending elicitation.

    roxy doesn't wait for a meaningful body — it just wants the handler to
    know so any server-side state tied to the flow can be cleaned up.
    """
    name: str = request.get("name", "")     # tool the elicitation belonged to
    action: str = request.get("action", "")  # 'decline' or 'cancel'
    logging.info("Elicitation %s for tool: %s", action, name)
    return {"ok": True}


def dispatch(request: Request) -> Reply:
    """Route an envelope to the matching handler based on ``type``."""
    request_type = request.get("type")

    if request_type == "discover":
        return handle_discover()
    if request_type == "call_tool":
        return handle_call_tool(request)
    if request_type == "read_resource":
        return handle_read_resource(request)
    if request_type == "get_prompt":
        return handle_get_prompt(request)
    if request_type == "elicitation_cancelled":
        return handle_elicitation_cancelled(request)

    return {
        "error": {
            "code": 400,
            "message": f"Unknown request type: {request_type}",
        },
    }


# ---------------------------------------------------------------------------
# HTTP server
# ---------------------------------------------------------------------------


class RoxyHandler(BaseHTTPRequestHandler):
    """Minimal POST-only HTTP handler that decodes JSON and dispatches."""

    # Silence the default per-request stderr logging; use the logging module
    # instead so the output plays well with roxy's own logs.
    def log_message(self, format: str, *args: Any) -> None:  # noqa: A002
        logging.debug("%s - %s", self.address_string(), format % args)

    def _send_json(self, status: int, payload: Reply) -> None:
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self) -> None:  # noqa: N802 — required name from BaseHTTPRequestHandler
        length = int(self.headers.get("Content-Length", "0") or "0")
        raw = self.rfile.read(length) if length > 0 else b""

        try:
            body = json.loads(raw.decode("utf-8")) if raw else None
        except (UnicodeDecodeError, json.JSONDecodeError):
            self._send_json(400, {"error": {"code": 400, "message": "Invalid JSON body"}})
            return

        if not isinstance(body, dict) or not isinstance(body.get("type"), str):
            self._send_json(
                400,
                {"error": {"code": 400, "message": "Invalid request: missing type field"}},
            )
            return

        try:
            reply = dispatch(body)
        except Exception as exc:  # noqa: BLE001 — any handler error becomes a 500
            logging.exception("handler failed")
            self._send_json(500, {"error": {"code": 500, "message": str(exc)}})
            return

        self._send_json(200, reply)

    def do_GET(self) -> None:  # noqa: N802
        self._send_json(405, {"error": {"code": 405, "message": "Method not allowed"}})


def main() -> None:
    logging.basicConfig(
        level=os.environ.get("LOG_LEVEL", "INFO"),
        format="%(asctime)s %(levelname)s %(message)s",
    )

    port = int(os.environ.get("PORT", "8000"))
    httpd = ThreadingHTTPServer(("127.0.0.1", port), RoxyHandler)
    logging.info("roxy example upstream listening on http://127.0.0.1:%d/", port)
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        logging.info("shutting down")
        httpd.server_close()


if __name__ == "__main__":
    main()
